std/assembly/finalizationregistry.ts

```diff
@@ -0,0 +1,76 @@
+// @ts-ignore: decorators
+@lazy let PREVIOUS_FINALIZER: i32 = 0;
+
+// @ts-ignore: decorators
+@lazy let INITIALIZED: boolean = false;
+
+// @ts-ignore: decorators
+@lazy let REGISTRIES_FOR_PTR: Map<usize, BaseRegistry[]> = new Map<usize, BaseRegistry[]>();
+
+abstract class BaseRegistry {
+  static finalizeAll(ptr: usize): void {
+    if (!REGISTRIES_FOR_PTR.has(ptr)) { return; }
+
+    const registries = REGISTRIES_FOR_PTR.get(ptr);
+    for (
+      let i = 0, numRegistries = registries.length;
+      i < numRegistries;
+      ++i
+    ) {
+      registries[i].finalize(ptr);
+    }
+    REGISTRIES_FOR_PTR.delete(ptr);
+
+    if (PREVIOUS_FINALIZER) {
+      call_indirect<void>(PREVIOUS_FINALIZER, ptr);
+    }
+  }
+
+  protected abstract finalize(ptr: usize): void;
+}
+
+export class FinalizationRegistry<T> extends BaseRegistry {
+  private entries: Map<usize, T> = new Map<usize, T>();
+
+  constructor(private finalizer: (heldValue: T) => void) {
+    super();
+
+    if (!INITIALIZED) {
+      PREVIOUS_FINALIZER = __finalize;
+      __finalize = BaseRegistry.finalizeAll.index;
+      INITIALIZED = true;
+    }
+  }
+
+  register<U>(key: U, value: T): void {
+    assert(isManaged<U>());
```

Comment:
Could this be a compile-time error so using an invalid key type won't compile?
```suggestion
    if (!isManaged<U>()) ERROR("invalid target");
```
The corresponding error in Node seems to be "Uncaught TypeError: FinalizationRegistry.prototype.register: invalid target".

Reply:
I see, I didn't know about `ERROR`